core/meta/src/TClass.cxx

```diff
@@ -5464,30 +5464,33 @@ TClass *TClass::LoadClassCustom(const char *requestedname, Bool_t silent)
 
 void TClass::LoadClassInfo() const
 {
-   R__LOCKGUARD(gInterpreterMutex);
+   // Return early if class info is already loaded or cannot be loaded
+   if (!fCanLoadClassInfo)
+      return;
 
-   // If another thread executed LoadClassInfo at about the same time
-   // as this thread return early since the work was done.
-   if (!fCanLoadClassInfo) return;
+   bool AutoParse = !gInterpreter->IsAutoParsingSuspended();
 
-   // If class info already loaded then do nothing.  This can happen if the
-   // class was registered by a dictionary, but the info came from reading
-   // the pch.
-   // Note: This check avoids using AutoParse for classes in the pch!
-   if (fClassInfo) {
-      return;
-   }
+   {
+      R__LOCKGUARD(gInterpreterMutex);
+
+      // Return early if another thread loaded the info while we were waiting for the lock
+      if (!fCanLoadClassInfo)
+         return;
+
+      if (AutoParse)
+         gInterpreter->AutoParse(GetName());
+
+      if (!fClassInfo)
+         gInterpreter->SetClassInfo(const_cast<TClass*>(this));
 
-   gInterpreter->AutoParse(GetName());
-   if (!fClassInfo) gInterpreter->SetClassInfo(const_cast<TClass*>(this));   // sets fClassInfo pointer
-   if (!gInterpreter->IsAutoParsingSuspended()) {
-      if (!fClassInfo) {
-         ::Error("TClass::LoadClassInfo",
-                 "no interpreter information for class %s is available even though it has a TClass initialization routine.",
-                 fName.Data());
-      }
       fCanLoadClassInfo = kFALSE;
```

Comment:
After SetClassInfo is done we can be in many state:
1) fClassInfo is now a value, we are all good, no new to try reloading (independently of the auto-parsing flag)
2) fClassInfo is still a nullptr but auto-parsing is enabled, there is no more information to be have anywhere, the AST decl for the class does not exist
3) fClassInfo is still anullptr but auto-parsing is disabled, we now know the AST decl for the class is not in the AST *but* it **might** be brought into the AST if we were to retry with auto-parsing on (because the header declaring it might be parsed).